### PR TITLE
Demographic cohort

### PR DESCRIPTION
**Pull request by shintoLamp**
Firstly, ignore the fact I called this `service-use-cohort` - I got confused and it is in fact the demographic cohort code.

There are four new functions, one for the main demographics, one for EOL cohort, one for substance misuse, and one for creating and writing out the demographic lookup.

Each of the first three functions have tests that I'm happy with, meaning that the final code doesn't really need one of its own. I have tested the final function against SLF temp file 6 and compared to the actual lookup in hscdiip, the only differences are those situations with F13 substance misuse cases being classified properly in the R output.

### CODE REVIEW

File: R/assign_demographic_cohort.R [hunk 1]
```diff
@@ -0,0 +1,108 @@
+#' Assign the demographic cohort variables
+#'
+#' @param data A data frame with the required variables for assignment
+#'
+#' @return A data frame with ten additional variables relating to the different cohorts
+#' @export
```
Comment (Jennit07) [inline R/assign_demographic_cohort.R:6]: Remove export to non top level functions

File: R/assign_demographic_cohort.R [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+      # Maternity classification
+      maternity_cohort = .data$recid == "02B",
+
+      # High CC classification
+      # FOR FUTURE: PhysicalandSensoryDisabilityClientGroup or LearningDisabilityClientGroup = "Y",
+      # then high_cc_cohort = TRUE
+      # FOR FUTURE: Care home removed, here's the code: .data$recid = "CH" & age < 65
+      high_cc_cohort =
+        purrr::reduce(dplyr::select(
```
Comment (Jennit07) [inline R/assign_demographic_cohort.R:69]: Use rowsums instead of purrr::reduce

File: R/assign_demographic_cohort.R (hunk 1 above)
Comment (shintoLamp) [reply R/assign_demographic_cohort.R]: Done

File: R/assign_demographic_cohort.R (hunk 2 above)
Comment (shintoLamp) [reply R/assign_demographic_cohort.R]: Used rowsums

File: R/create_demographic_lookup.R [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+#' @export
+#'
+#' @seealso \itemize{\item[assign_demographic_cohort()]
+#'                   \item[assign_eol_cohort()]
+#'                   \item[assign_substance_cohort()]
+#'                   \item[get_cohorts_paths()]}
+#'
+#' @family Demographic and Service Use Cohort functions
+create_demographic_lookup <- function(data, year, write_to_disk = TRUE) {
+  demo_lookup <- data %>%
```
[the remark this answers is not in the document]
Comment (Moohan) [reply R/create_demographic_lookup.R:15]: I think the `check_variables_exist` should be brought out of the 'lower level' functions and put here (or possibly left in both places as it's a quick function to run). I've had the same issues trying to test it.

File: R/create_demographic_lookup.R [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    "deathdiag11"
+  ))
+
+  external_deaths <- c(
+    glue::glue("V{stringr::str_pad(1:99, 2, 'left', '0')}"),
+    glue::glue("W{stringr::str_pad(0:99, 2, 'left', '0')}"),
+    glue::glue("X{stringr::str_pad(0:99, 2, 'left', '0')}"),
+    glue::glue("Y{stringr::str_pad(0:84, 2, 'left', '0')}")
+  )
+  falls_codes <- c(glue::glue("W{stringr::str_pad(0:19, 2, 'left', '0')}"))
```
Comment (Moohan) [inline R/create_demographic_lookup.R:205]: Add comments for these to show what the code should be generating e.g. V01-V99 etc.

File: R/create_demographic_lookup.R [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+#'
+#' @family Demographic and Service Use Cohort functions
+create_demographic_lookup <- function(data, year, write_to_disk = TRUE) {
+  demo_lookup <- data %>%
+    # Remove missing chi
+    dplyr::filter(!is_missing(.data$chi)) %>%
+    # Add the various cohorts
+    assign_demographic_cohort() %>%
+    assign_eol_cohort() %>%
+    assign_substance_cohort() %>%
```
[the remark this answers is not in the document]
Comment (Moohan) [reply R/create_demographic_lookup.R:21]: Having read the full PR, I think a function per cohort might be a nice way to do this?

**Review by Moohan — COMMENTED** [review]
I think there's still a bit of work to do to make the code as readable and maintainable as possible.

I was also doing some checks between the old SPSS and the new SPSS (with the F11 to F13 correction) and it wasn't producing any differences. It's possible I have done something wrong in the SPSS but I suspect that the difference we're seeing are to do with this code.

To test we should compare the output of this to the output from the SPSS - All of the individual cohorts should be exactly the same except for substance misuse. Further, most of the 'final' demographic cohorts should be the same, except for those affected by the change to their substance misuse cohort (i.e. some who were in a 'healthier' cohort before might now be in substance misuse).

Once you've had a look at the changes we can discuss how to check it against the SPSS output if you want?

I think most of my comments here will also apply to #414

File: R/create_demographic_lookup.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (shintoLamp) [reply R/create_demographic_lookup.R:15]: Added `check_variables_exist` to the main function, with the additional variables needed to do the whole thing

File: R/create_demographic_lookup.R [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    # dtplyr::lazy_dt() %>%
+    dplyr::group_by(.data$chi, .data$cij_marker) %>%
+    dplyr::summarise(
+      dplyr::across(c(
+        dplyr::contains("cohort"),
+        .data$f11, .data$t402_t404, .data$f13, .data$t424
+      ), any)
+    ) %>%
+    dplyr::ungroup() %>%
+    # tibble::as_tibble() %>%
```
[the remark this answers is not in the document]
Comment (shintoLamp) [reply R/create_demographic_lookup.R:33]: For some reason I can't get the dtplyr solution to work here so I've removed it

File: R/create_demographic_lookup.R [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+      (.data$f11 & .data$t402_t404) | (.data$f13 & .data$t424), TRUE, .data$substance_cohort
+    )) %>%
+    # Aggregate to CHI level
+    # dtplyr::lazy_dt() %>%
+    dplyr::group_by(.data$chi) %>%
+    dplyr::summarise(dplyr::across(c(dplyr::contains("cohort")), any)) %>%
+    dplyr::ungroup() %>%
+    # tibble::as_tibble() %>%
+    # Rename variables
+    dplyr::rename_with(~ stringr::str_sub(.x, end = -8), dplyr::ends_with("_cohort")) %>%
```
[the remark this answers is not in the document]
Comment (shintoLamp) [reply R/create_demographic_lookup.R:45]: Done

File: R/create_demographic_lookup.R [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    dplyr::relocate(.data$demographic_cohort, .after = chi)
+
+  # Write to disk
+  if (write_to_disk == TRUE) {
+    write_rds(demo_lookup,
+      path = glue::glue("{get_slf_dir()}/Cohorts/Demographic_Cohorts_{year}.rds")
+    )
+  } else {
+    return(demo_lookup)
+  }
```
[the remark this answers is not in the document]
Comment (shintoLamp) [reply R/create_demographic_lookup.R:71]: Done

File: R/create_demographic_lookup.R (hunk 4 above)
Comment (shintoLamp) [reply R/create_demographic_lookup.R:205]: Done

File: R/create_demographic_lookup.R [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+#'    \item The significant facility is 1E or 1D
+#'    \item The recid is GLS}}}
+#'
+#' @param recid A vector of record IDs
+#' @param diag1 A vector of Diagnosis Code 1
+#' @param diag2 A vector of Diagnosis Code 2
+#' @param diag3 A vector of Diagnosis Code 3
+#' @param diag4 A vector of Diagnosis Code 4
+#' @param diag5 A vector of Diagnosis Code 5
+#' @param diag6 A vector of Diagnosis Code 6
```
Comment (Moohan) [inline R/create_demographic_lookup.R:142]: See https://cran.r-project.org/web/packages/roxygen2/vignettes/reuse.html specifically
```suggestion
#' @inheritParams assign_mh_cohort 
```
Which would cut down a lot of lines and make it slightly more maintainable.

I think we could probably make much better use of documentation inheritance throughout the package tbh but this seemed a particularly useful case for it.

File: R/create_demographic_lookup.R [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+          ~ stringr::str_sub(.x, 1, 2) %in%
+            c("F2", "F3")
+        )) > 0 |
+          rowSums(dplyr::across(
+            c("diag1", "diag2", "diag3", "diag4", "diag5", "diag6"),
+            ~ stringr::str_sub(.x, 1, 4) %in%
+              c("F067", "F070", "F072", "F078", "F079")
+          )) > 0))
+  return(mh)
+}
```
Comment (Moohan) [inline R/create_demographic_lookup.R:123]: This whole thing was going quite slowly. This was the best I could do to speed it up, which can be applied to most / all the other functions. It's about a 30% speedup, I think its doing the `rowSums` + `across` bit once which makes the most difference.

```suggestion
assign_mh_cohort <- function(recid, diag1, diag2, diag3, diag4, diag5, diag6) {
  mh <-
    # FOR FUTURE: when variable MentalHealthProblemsClientGroup exists and is "Y", mh_cohort = TRUE
    dplyr::case_when(
    recid == "04B" ~ TRUE,
    recid %in% c("01B", "GLS", "50B", "02B", "AE2") &
      (rowSums(dplyr::across(
        c("diag1", "diag2", "diag3", "diag4", "diag5", "diag6"),
        ~ stringr::str_starts(.x, paste("F2", "F3", "F067", "F070", "F072", "F078", "F079", sep = "|")))) > 0) ~ TRUE,
      TRUE ~ FALSE
  )
  
  return(mh)
}
```

File: R/create_demographic_lookup.R [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+      comm_living ~ "Assisted Living in the Community",
+      TRUE ~ "Healthy and Low User"
+    )) %>%
+    # Reorder variables
+    dplyr::relocate(.data$demographic_cohort, .after = chi)
+
+  # Write to disk
+  if (write_to_disk == TRUE) {
+    write_rds(demo_lookup,
+      path = glue::glue("{get_slf_dir()}/Cohorts/Demographic_Cohorts_{year}.rds")
```
Comment (Moohan) [inline R/create_demographic_lookup.R:85]: Create a function for this (unless it already exists - possibly make existing rds)

File: R/create_demographic_lookup.R [hunk 12]
```diff
@@ -0,0 +1,392 @@
+#' Create the demographic lookup file
+#'
+#' @param data A data frame with the required variables
+#' @param year The year in standard SLF format
+#' @param write_to_disk Defaults to TRUE
+#'
+#' @export
+#'
+#' @family Demographic and Service Use Cohort functions
+create_demographic_lookup <- function(data, year, write_to_disk = TRUE) {
```
Comment (Jennit07) [inline R/create_demographic_lookup.R:10]:
```suggestion
create_demographic_cohort <- function(data, year, write_to_disk = TRUE) {
```
One small name change suggestion as i just noticed this and we might get confused with the social care demographic lookup

File: R/create_demographic_lookup.R [hunk 13]
```diff
@@ -0,0 +1,392 @@
+#' Create the demographic lookup file
```
Comment (Jennit07) [inline R/create_demographic_lookup.R:1]:
```suggestion
#' Create the demographic cohort file
```

File: R/create_demographic_lookup.R (hunk 11 above)
Comment (Jennit07) [reply R/create_demographic_lookup.R:81]: function already called `get_demog_cohorts_path` in rds format

File: R/create_demographic_lookup.R (hunk 12 above)
Comment (Moohan) [reply R/create_demographic_lookup.R:10]:
```suggestion
create_demographic_cohorts <- function(data, year, write_to_disk = TRUE) {
```

File: R/create_demographic_lookup.R (hunk 13 above)
Comment (Moohan) [reply R/create_demographic_lookup.R:1]:
```suggestion
#' Create the demographic cohorts lookup file
```

File: R/get_cohorts_paths.R [hunk 14]
```diff
@@ -1,34 +1,34 @@
-#' Demographics Cohort File Path
+#' Demographic cohort lookup Path
```
Comment (Moohan) [inline R/get_cohorts_paths.R:1]:
```suggestion
#' Demographic cohorts lookup Path
```

File: R/get_cohorts_paths.R [hunk 15]
```diff
@@ -1,34 +1,34 @@
...
 #'
-#' @description Get the full path to the Demographics Cohort lookup
+#' @description Get the path to the demographic cohort lookup, there is one
```
Comment (Moohan) [inline R/get_cohorts_paths.R:3]:
```suggestion
#' @description Get the path to the demographic cohorts lookup, there is one
```

File: R/get_cohorts_paths.R [hunk 16]
```diff
@@ -1,34 +1,34 @@
...
+    file_name = glue::glue("demographic_cohorts_{year}.rds"),
     ...
   )
 
-  return(demog_cohorts_path)
+  return(demographic_cohorts_path)
 }
 
-#' Service Use Cohort File Path
+#' Service-use cohort lookup Path
```
Comment (Moohan) [inline R/get_cohorts_paths.R:23]:
```suggestion
#' Service-use cohorts lookup Path
```

File: R/get_cohorts_paths.R [hunk 17]
```diff
@@ -1,34 +1,34 @@
...
 #'
-#' @description Get the full path to the Service Use Cohort lookup
+#' @description Get the path to the service-use cohort lookup, there is one
```
Comment (Moohan) [inline R/get_cohorts_paths.R:25]:
```suggestion
#' @description Get the path to the service-use cohorts lookup, there is one
```

File: R/get_cohorts_paths.R [hunk 18]
```diff
@@ -1,34 +1,34 @@
...
+#' @description Get the path to the demographic cohort lookup, there is one
+#' lookup per year.
 #'
-#' @param year The year for the cohorts extract
+#' @param year financial year in '1718' format
 #' @param ... additional arguments passed to [get_file_path()]
 #'
-#' @return The path to the Demographics Cohort lookup
-#' as a [fs::path()]
+#' @return The path to the demographic lookup as an [fs::path()]
```
Comment (Moohan) [inline R/get_cohorts_paths.R:9]:
```suggestion
#' @return The path to the demographic cohorts lookup as an [fs::path()]
```

File: R/get_cohorts_paths.R [hunk 19]
```diff
@@ -1,34 +1,34 @@
...
+#' @description Get the path to the service-use cohort lookup, there is one
+#' lookup per year.
 #'
-#' @param year The year for the cohorts extract
+#' @param year financial year in '1718' format
 #' @param ... additional arguments passed to [get_file_path()]
 #'
-#' @return The path to the Service Use Cohort lookup
-#' as a [fs::path()]
+#' @return The path to the service-use lookup as an [fs::path()]
```
Comment (Moohan) [inline R/get_cohorts_paths.R:31]:
```suggestion
#' @return The path to the service-use cohorts lookup as an [fs::path()]
```

**Review by Moohan — APPROVED** [review]
I think I'm happy with this now 👍 

@shintoLamp please merge once you've seen the last few changes. @Jennit07 I implemented your suggested name changes, so I assume you're also happy now?

File: R/get_cohorts_paths.R [hunk 20]
```diff
@@ -1,34 +1,34 @@
...
-#' @family cohort lookup file paths
+#' @family cohorts file path
 #' @seealso [get_file_path()] for the generic function.
-get_demog_cohorts_path <- function(year, ...) {
-  demog_cohorts_path <- get_file_path(
+get_demographic_cohorts_path <- function(year, ...) {
+  demographic_cohorts_path <- get_file_path(
     directory = fs::path(get_slf_dir(), "Cohorts"),
-    file_name = glue::glue("Demographic_Cohorts_{year}.rds"),
+    file_name = glue::glue("demographic_cohorts_{year}.rds"),
```
Comment (Jennit07) [inline R/get_cohorts_paths.R:16]: Due to this change to lowercase, there is one test now failing as the file 'does not exist' (due to capitalization in the folder)

**Review by Jennit07 — APPROVED** [review]
Im going to approve this as this looks good to me. One comment which needs addressed before merging is that due to changing the demographic cohort file name to lower case, this is causing one test to fail. Very minor so going to approve this for now

File: R/get_cohorts_paths.R (hunk 20 above)
Comment (Moohan) [reply R/get_cohorts_paths.R:16]: Argh, that's my fault as I thought we could change the naming of the R files without affecting any SPSS things. We could probably just drop the test which looks for the zsav file (we will at some anyway!)

**Pull request merged**